src/Cake.Testing/Fakes/FakeFileSystemTree.cs

```diff
@@ -0,0 +1,245 @@
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Cake.Core;
+using Cake.Core.IO;
+using Path = Cake.Core.IO.Path;
+
+namespace Cake.Testing.Fakes
+{
+    internal sealed class FakeFileSystemTree
+    {
+        private readonly PathComparer _comparer;
+        private readonly FakeDirectory _root;
+
+        public PathComparer Comparer
+        {
+            get { return _comparer; }
+        }
+
+        public FakeFileSystemTree(ICakeEnvironment environment)
+        {
+            if (environment == null)
+            {
+                throw new ArgumentNullException("environment");
+            }
+            if (environment.WorkingDirectory == null)
+            {
+                throw new ArgumentException("Working directory not set.");
+            }
+            if (environment.WorkingDirectory.IsRelative)
+            {
+                throw new ArgumentException("Working directory cannot be relative.");
+            }
+            _comparer = new PathComparer(environment.IsUnix());
+
+            _root = new FakeDirectory(this, "/");
+            _root.Create();
+        }
+
+        public FakeDirectory CreateDirectory(DirectoryPath path)
+        {
+            return CreateDirectory(new FakeDirectory(this, path));
+        }
+
+        public FakeDirectory CreateDirectory(FakeDirectory directory)
+        {
+            var path = directory.Path;
+            var queue = GetSegmentQueue(path);
+
+            FakeDirectory current = null;
+            var children = _root.Content;
+
+            while (queue.Count > 0)
+            {
+                // Get the segment.
+                var currentSegment = queue.Dequeue();
+                var parent = current;
+
+                // Calculate the current path.
+                path = parent != null ? parent.Path.Combine(currentSegment) : new DirectoryPath(currentSegment);
+
+                if (!children.Directories.ContainsKey(path))
+                {
+                    if (queue.Count == 0)
+                    {
+                        // Use the provided directory.
+                        current = directory;
+                    }
+                    else
+                    {
+                        current = new FakeDirectory(this, path);
+                    }
+
+                    current.Parent = parent;
+                    current.Hidden = false;
+                    children.Add(current);
+                }
+                else
+                {
+                    current = children.Directories[path];
+                }
+
+                current.Exists = true;
+                children = current.Content;
+            }
+
+            return directory;
+        }
+
+        public void DeleteDirectory(FakeDirectory fakeDirectory, bool recursive)
+        {
+            var root = new Stack<FakeDirectory>();
+            var result = new Stack<FakeDirectory>();
+
+            root.Push(fakeDirectory);
```

Comment:
Should make sure that `fakeDirectory` really exist before pushing it on the root.